File: src/fensu/mapping/_helpers/render.py
"""Render call-map trees as deterministic terminal text."""

from __future__ import annotations

from pathlib import Path

from fensu.mapping.constants import (
    ANSI_CYCLE,
    ANSI_DIM,
    ANSI_FUNCTION,
    ANSI_RESET,
    ANSI_UNRESOLVED,
    MAX_COMPACT_PATH_PARTS,
)
from fensu.mapping.models import CallMapNode, UnresolvedCall
from fensu.mapping.types import PathMode


def render_tree(
    *,
    root: CallMapNode,
    repo_root: Path,
    path_mode: PathMode = PathMode.RELATIVE,
    use_color: bool = False,
) -> str:
    """Render one call tree with locations, cycle markers, and depth markers."""

    child_lines: tuple[str, ...] = _child_lines(
        node=root,
        repo_root=repo_root,
        path_mode=path_mode,
        prefix="",
        use_color=use_color,
    )
    root_label: str = _label(
        node=root, repo_root=repo_root, path_mode=path_mode, use_color=use_color
    )
    return "\n".join((root_label, *child_lines)) + "\n"


def _child_lines(
    *,
    node: CallMapNode,
    repo_root: Path,
    path_mode: PathMode,
    prefix: str,
    use_color: bool,
) -> tuple[str, ...]:
    lines: list[str] = []
    for index, entry in enumerate(node.entries):
        last: bool = index == len(node.entries) - 1
        connector: str = "└── " if last else "├── "
        rendered_connector: str = _color(
            text=f"{prefix}{connector}", style=ANSI_DIM, enabled=use_color
        )
        if isinstance(entry, UnresolvedCall):
            location: str = _location(
                path=node.definition.path,
                line=entry.line,
                repo_root=repo_root,
                path_mode=path_mode,
            )
            label: str = _color(text=f"{entry.name}(...)", style=ANSI_FUNCTION, enabled=use_color)
            rendered_location: str = _color(text=location, style=ANSI_DIM, enabled=use_color)
            marker: str = _color(
                text=f"(unresolved {entry.reason})",
                style=ANSI_UNRESOLVED,
                enabled=use_color,
            )
            lines.append(f"{rendered_connector}{label}{rendered_location}  {marker}")
            continue
        label = _label(
            node=entry,
            repo_root=repo_root,
            path_mode=path_mode,
            use_color=use_color,
        )
        lines.append(f"{rendered_connector}{label}")
        child_prefix: str = f"{prefix}{'    ' if last else '│   '}"
        lines.extend(
            _child_lines(
                node=entry,
                repo_root=repo_root,
                path_mode=path_mode,
                prefix=child_prefix,
                use_color=use_color,
            )
        )
    return tuple(lines)


def _label(*, node: CallMapNode, repo_root: Path, path_mode: PathMode, use_color: bool) -> str:
    display_name: str = (
        f"{node.dispatch_class_name}.{node.definition.name}"
        if node.dispatch_class_name is not None
        else node.definition.display_name
    )
    function: str = _color(text=f"{display_name}(...)", style=ANSI_FUNCTION, enabled=use_color)
    location: str = _location(
        path=node.definition.path,
        line=node.definition.syntax.line,
        repo_root=repo_root,
        path_mode=path_mode,
    )
    rendered_location: str = _color(text=location, style=ANSI_DIM, enabled=use_color)
    if node.cycle:
        marker: str = _color(text="  (cycle)", style=ANSI_CYCLE, enabled=use_color)
    elif node.truncated:
        marker = _color(text="  (depth limit)", style=ANSI_DIM, enabled=use_color)
    else:
        marker = ""
    return f"{function}{rendered_location}{marker}"


def _location(*, path: Path, line: int, repo_root: Path, path_mode: PathMode) -> str:
    if path_mode is PathMode.NONE:
        return ""
    display_path: Path = path
    if path_mode is not PathMode.ABSOLUTE and path.is_relative_to(repo_root):
        display_path = path.relative_to(repo_root)
    path_text: str = display_path.as_posix()
    if path_mode is PathMode.COMPACT:
        path_text = _compact_path(path_text)
    return f"  {path_text}:{line}"


def _compact_path(path: str) -> str:
    parts: tuple[str, ...] = tuple(path.split("/"))
    if len(parts) <= MAX_COMPACT_PATH_PARTS:
        return path
    return "/".join((*parts[:2], "…", *parts[-2:]))


def _color(*, text: str, style: str, enabled: bool) -> str:
    if not text or not enabled:
        return text
    return f"{style}{text}{ANSI_RESET}"

File: crates/fensu-cli/src/mapping/helpers/render.rs
use std::path::Path;

use crate::mapping::constants::MAX_COMPACT_PATH_PARTS;
use crate::mapping::models::{CallMapEntry, CallMapNode, PathMode};

const RESET: &str = "\x1b[0m";
const DIM: &str = "\x1b[2m";
const FUNCTION: &str = "\x1b[1;36m";
const UNRESOLVED: &str = "\x1b[33m";
const CYCLE: &str = "\x1b[35m";

pub(crate) fn render(
    root: &CallMapNode,
    repo_root: &Path,
    path_mode: PathMode,
    use_color: bool,
) -> String {
    let mut lines = vec![label(root, repo_root, path_mode, use_color)];
    child_lines(root, repo_root, path_mode, "", use_color, &mut lines);
    format!("{}\n", lines.join("\n"))
}

fn child_lines(
    node: &CallMapNode,
    repo_root: &Path,
    path_mode: PathMode,
    prefix: &str,
    use_color: bool,
    lines: &mut Vec<String>,
) {
    for (position, entry) in node.entries.iter().enumerate() {
        let last = position + 1 == node.entries.len();
        let connector = if last { "└── " } else { "├── " };
        let rendered_connector = color(&format!("{prefix}{connector}"), DIM, use_color);
        match entry {
            CallMapEntry::Unresolved(unresolved) => {
                let location =
                    location(&node.definition.path, unresolved.line, repo_root, path_mode);
                let function = color(&format!("{}(...)", unresolved.name), FUNCTION, use_color);
                let rendered_location = color(&location, DIM, use_color);
                let marker = color(
                    &format!("(unresolved {})", unresolved.reason),
                    UNRESOLVED,
                    use_color,
                );
                lines.push(format!(
                    "{rendered_connector}{function}{rendered_location}  {marker}"
                ));
            }
            CallMapEntry::Node(child) => {
                lines.push(format!(
                    "{rendered_connector}{}",
                    label(child, repo_root, path_mode, use_color)
                ));
                let child_prefix = format!("{prefix}{}", if last { "    " } else { "│   " });
                child_lines(child, repo_root, path_mode, &child_prefix, use_color, lines);
            }
        }
    }
}

fn label(node: &CallMapNode, repo_root: &Path, path_mode: PathMode, use_color: bool) -> String {
    let display_name = node.dispatch_class_name.as_ref().map_or_else(
        || node.definition.qualified_name(),
        |class| format!("{class}.{}", node.definition.name),
    );
    let function = color(&format!("{display_name}(...)"), FUNCTION, use_color);
    let location = location(
        &node.definition.path,
        node.definition.syntax.line,
        repo_root,
        path_mode,
    );
    let rendered_location = color(&location, DIM, use_color);
    let marker = if node.cycle {
        color("  (cycle)", CYCLE, use_color)
    } else if node.truncated {
        color("  (depth limit)", DIM, use_color)
    } else {
        String::new()
    };
    format!("{function}{rendered_location}{marker}")
}

fn location(path: &Path, line: u32, repo_root: &Path, path_mode: PathMode) -> String {
    if path_mode == PathMode::None {
        return String::new();
    }
    let display = if path_mode != PathMode::Absolute {
        path.strip_prefix(repo_root).unwrap_or(path)
    } else {
        path
    };
    let mut text = display.to_string_lossy().replace('\\', "/");
    if path_mode == PathMode::Compact {
        let parts = text.split('/').collect::<Vec<_>>();
        if parts.len() > MAX_COMPACT_PATH_PARTS {
            text = format!(
                "{}/{}/…/{}/{}",
                parts[0],
                parts[1],
                parts[parts.len() - 2],
                parts[parts.len() - 1]
            );
        }
    }
    format!("  {text}:{line}")
}

fn color(text: &str, style: &str, enabled: bool) -> String {
    if text.is_empty() || !enabled {
        text.to_owned()
    } else {
        format!("{style}{text}{RESET}")
    }
}
